### PR TITLE
Added optimizations for combining conditions together when their resu…

### PR DESCRIPTION
**Pull request by ghost**
…lts are OR'ed

Added recognition of impossible comparisons via NZM and optimize them away
Recognize (x + -y) and transform to (x - y) for constants
Recognize (~x ) + 1 and transform to -x
Check and transform comparisons if theyre semantically equal to others
Detect comparisons of single-bit values with their only possible non-zero value and transform to true/false tests
Transform ==0 to IS_FALSE, !=0 to IS_TRUE
Truncate to int8 if operand for IS_TRUE/IS_FALSE has a nzm of 1
Reduced code generated for SubDidCarry slightly
Add special case for InstrEmit_srawix if mask == 1
Cut down the code generated for trap instructions, instead of naive or'ing or compare results do a switch and select the best condition
Rerun simplification pass until no changes, as some optimizations will enable others to be done
Enable rel32 call optimization by default

### CODE REVIEW

File: src/xenia/cpu/hir/instr.h [hunk 1]
```diff
@@ -59,7 +59,6 @@ class Instr {
   void MoveBefore(Instr* other);
   void Replace(const OpcodeInfo* new_opcode, uint16_t new_flags);
   void Remove();
-
```
Comment (Margen67) [inline src/xenia/cpu/hir/instr.h:62]: Is this intentional?

File: src/xenia/cpu/hir/instr.h [hunk 2]
```diff
@@ -105,6 +104,17 @@ if both are constant, return nullptr, nullptr
   }
 
   Instr* GetDestDefSkipAssigns();
+  // returns [def op, constant]
+  std::pair<Value*, Value*> BinaryValueArrangeByDefOpAndConstant(
+      const OpcodeInfo* op_ptr) {
+    auto result = BinaryValueArrangeByDefiningOpcode(op_ptr);
+
+    if (!result.first) return result;
```
Comment (Gliniak) [inline src/xenia/cpu/hir/instr.h:112]: Styling

File: src/xenia/cpu/hir/instr.h [hunk 3]
```diff
@@ -105,6 +104,17 @@ if both are constant, return nullptr, nullptr
...
+    if (!result.second->IsConstant()) {
```
Comment (Gliniak) [inline src/xenia/cpu/hir/instr.h:113]: Is there a chance that second can be null? What then?

Comment (ghost) [reply src/xenia/cpu/hir/instr.h:113]: nope, it only returns NULL, NULL

**Pull request merged**